cmd/zfs/zfs_main.c

```diff
@@ -64,6 +64,8 @@
 #include <sys/types.h>
 #include <time.h>
 #include <sys/zfs_project.h>
+#include <sys/spa.h>
+#include <sys/zio.h>
```

Comment:
We want to avoid including `sys/spa.h` and `sys/zio.h` in the command line tools.  It looks like somewhere along the line they were added to `libzfs_impl.h` to pick up a few handy `#define`s like `SPA_MINBLOCKSIZE` and `ZIO_CRYPT_OFF`.  But none of the functions in these headers are available to the utilities or libzfs.  As part of this fix it seems like we should remove these includes from `libzfs_impl.h` and relocate any of these shared `#define`s to a header which is intended to be included in both the user and kernel space, such as `include/sys/fs/zfs.h`.  Or alternately, just don't use the macro for these values which can never really change.

Reply:
But `spa.h` is in `COMMON_H`?

Reply:
Yes, I can see how that would be confusing.  Everything in `COMMON_H` is needed to build the kmods in kernel space, _and_ the `libzpool.so` library.  `libzpool.so` basically provides a full user space build of the kernel modules as a library.  It's used by `ztest(1)` to stress test the core filesystem code (VDEVS, SPA, DMU, etc) in user space for easier testing/development.  It's also used by `zdb(8)` in order to effectively import a pool read-only in user space for debugging and digging through the on-disk format.

The user space CLI utilities (zfs, zpool, zinject) shouldn't depend on anything in `libzpool.so` or its headers. They depend on  `libzfs.so` and `libzfs_core.so` which provide the user space interfaces for interacting with the kernel modules via ioctls.  There are some places where the line gets a little blurry though and it's tempting "borrow" something from common headers so it can be used in one of the CLI utilities.  Because it was convenient is almost certainly how they ended up being included in `libzfs_impl.h`.

Reply:
Ah, makes sense, thanks. I've moved the chunks with SPA_MINBLOCKSHIFT and zio_encrypt into `sys/fs/zfs.h`, seemed least intrusive in the whole scheme of things.

Reply:
That strikes me as a pretty reasonable way to go as well.  Looks good.